pandas/io/xml.py

```diff
@@ -0,0 +1,966 @@
+                dicts = [
+                    {
+                        **el.attrib,
+                        **(
+                            {el.tag: el.text.strip()}
+                            if el.text and not el.text.isspace()
+                            else {}
+                        ),
+                        **{
+                            ch.tag: ch.text.strip() if ch.text else None
+                            for ch in el.findall("*")
+                        },
+                    }
+                    for el in elems
+                ]
+
+        dicts = [
+            {k.split("}")[1] if "}" in k else k: v for k, v in d.items()} for d in dicts
+        ]
+
+        keys = list(dict.fromkeys([k for d in dicts for k in d.keys()]))
+        dicts = [{k: d[k] if k in d.keys() else None for k in keys} for d in dicts]
+
+        if self.names:
+            dicts = [
+                {nm: v for nm, (k, v) in zip(self.names, d.items())} for d in dicts
+            ]
+
+        return dicts
+
+    def _validate_path(self) -> None:
+        """
+        Notes
+        -----
+        `etree` supports limited XPath. If user attempts a more complex
+        expression syntax error will raise.
+        """
+
+        msg = (
+            "xpath does not return any nodes. "
+            "If document uses namespaces denoted with "
+            "xmlns, be sure to define namespaces and "
+            "use them in xpath."
+        )
+        try:
+            elems = self.xml_doc.find(self.xpath, namespaces=self.namespaces)
+            if elems is None:
+                raise ValueError(msg)
+
+            if elems is not None and elems.find("*") is None and elems.attrib is None:
+                raise ValueError(msg)
+
+        except (KeyError, SyntaxError):
+            raise SyntaxError(
+                "You have used an incorrect or unsupported XPath "
+                "expression for etree library or you used an "
+                "undeclared namespace prefix."
+            )
+
+    def _validate_names(self) -> None:
+        if self.names:
+            parent = self.xml_doc.find(self.xpath, namespaces=self.namespaces)
+            if parent:
+                children = parent.findall("*")
+
+            if is_list_like(self.names):
+                if len(self.names) < len(children):
+                    raise ValueError(
+                        "names does not match length of child elements in xpath."
+                    )
+            else:
+                raise TypeError(
+                    f"{type(self.names).__name__} is not a valid type for names"
+                )
+
+    def _parse_doc(self) -> Union[Element, ElementTree]:
+        from xml.etree.ElementTree import (
+            Element,
+            ElementTree,
+            XMLParser,
+            fromstring,
+            parse,
+        )
+
+        if isinstance(self.path_or_buffer, str):
+            if self.path_or_buffer.startswith(("<?xml", "<")):
```

Comment:
Can combine the two `if`s. If you use the `isinstance` call multiple times, it might be nice to out source it to io/common (similar to `is_url`).

Reply:
Any change in pandas.io.common may need a different GH issue/PR. Also, in pandas.io.formats.xml, I had to use `# type: ignore[arg-type]` for `get_handles.handle.write` since for `is_text` it implements TextIOWrapper which inherits from [TextIOBase](https://docs.python.org/3/library/io.html#io.TextIOBase) so does not support `write(bytes)`. See [L692](https://github.com/ParfaitG/pandas/blob/read_xml/pandas/io/common.py#L692). Ok for csv and json but not xml. Maybe add `is_bytes` arg for `get_handle`?

Reply:
it would be cool if `get_handle` returns either `Buffer[str]` or `Buffer[bytes]` depending on whether `is_text` and `is_bytes` is used